crates/sui/src/client_commands.rs

```diff
@@ -840,6 +851,47 @@ impl SuiClientCommands {
                 };
                 SuiClientCommandResult::Addresses(output)
             }
+            SuiClientCommands::Balance {
+                address,
+                with_coins,
+            } => {
+                let address = get_identity_address(address, context)?;
+                let client = context.get_client().await?;
+                let mut objects: Vec<Coin> = Vec::new();
+                let mut cursor = None;
+                loop {
+                    let response = client
+                        .coin_read_api()
+                        .get_all_coins(address, cursor, None)
+                        .await?;
+                    objects.extend(response.data);
+
+                    if response.has_next_page {
+                        cursor = response.next_cursor;
+                    } else {
+                        break;
+                    }
+                }
+                let mut coins_by_type: BTreeMap<String, (Option<SuiCoinMetadata>, Vec<Coin>)> =
+                    BTreeMap::new();
+                for c in objects {
+                    if let Some(coins) = coins_by_type.get_mut(&c.coin_type) {
+                        coins.1.push(c);
+                    } else {
+                        let coin_metadata = client
+                            .coin_read_api()
+                            .get_coin_metadata(c.coin_type.clone())
+                            .await
+                            .map_err(|_| {
+                                anyhow!("Cannot fetch the coin metadata for coin {}", c.coin_type)
+                            })?;
+
+                        coins_by_type.insert(c.coin_type.clone(), (coin_metadata, vec![c]));
+                    }
```

Comment:
- Use the `BTreeMap::entry` API instead of `get_mut` + `insert` if you can. 
- Avoid discarding the error with `map_err(|_| { ... })`, instead use `with_context` and allow `anyhow` to combine the underlying error and your context for you.

Something like:

```rust
let coins = match coins_by_type.entry(c.coin_type.clone()) {
    Entry::Occupied(e) => e.get_mut().1,
    Entry::Vacant(e) => {
        let coin_metadata = client
            .coin_read_api()
            .get_coin_metadata(c.coin_type.clone())
            .await
            .with_context(|| format!("Cannot fetch the coin metadata for coin {}", c.coin_type))?;

        e.insert((coin_metadata, vec![])).1
    },
}

coins.insert(c);
```

Reply:
I ended up with this which is okish, but it still requires some cloning :(
```rust
                    let coin_type = c.coin_type.clone();
                    coins_by_type
                        .entry(coin_type.clone())
                        .and_modify(|val| val.1.push(c.clone()))
                        .or_insert({
                            let coin_metadata = client
                                .coin_read_api()
                                .get_coin_metadata(coin_type.clone())
                                .await
                                .with_context(|| {
                                    format!(
                                        "Cannot fetch the coin metadata for coin {}",
                                        coin_type.clone()
                                    )
                                })?;

                            (coin_metadata, vec![c])
                        });
```